commands/volumes/volume-option.go

```diff
@@ -0,0 +1,138 @@
+package volumecommands
+
+import (
+	"net/http"
+
+	"github.com/gluster/glusterd2/errors"
+	"github.com/gluster/glusterd2/gdctx"
+	"github.com/gluster/glusterd2/peer"
+	restutils "github.com/gluster/glusterd2/servers/rest/utils"
+	"github.com/gluster/glusterd2/transaction"
+	"github.com/gluster/glusterd2/utils"
+	"github.com/gluster/glusterd2/volume"
+	"github.com/pborman/uuid"
+
+	"github.com/gorilla/mux"
+)
+
+// VolOptionRequest represents an incoming request to set volume options
+type VolOptionRequest struct {
+	Options map[string]string `json:"options"`
+}
+
+func updateVolinfoOnOptionChange(c transaction.TxnCtx) error {
+
+	var volinfo volume.Volinfo
+	if err := c.Get("volinfo", &volinfo); err != nil {
+		return err
+	}
+
+	if err := volume.AddOrUpdateVolumeFunc(&volinfo); err != nil {
+		c.Logger().WithError(err).WithField(
+			"volume", volinfo.Name).Debug("failed to store volume info")
+		return err
+	}
+
+	return nil
+}
+
+func registerVolOptionStepFuncs() {
+	var sfs = []struct {
+		name string
+		sf   transaction.StepFunc
+	}{
+		{"vol-option.UpdateVolinfo", updateVolinfoOnOptionChange},
+		{"vol-option.RegenerateVolfiles", generateVolfiles},
+		{"vol-option.NotifyVolfileChange", notifyVolfileChange},
+	}
+	for _, sf := range sfs {
+		transaction.RegisterStepFunc(sf.sf, sf.name)
+	}
+}
+
+func volumeOptionsHandler(w http.ResponseWriter, r *http.Request) {
+
+	p := mux.Vars(r)
+	volname := p["volname"]
+	reqID, logger := restutils.GetReqIDandLogger(r)
+
+	volinfo, err := volume.GetVolume(volname)
+	if err != nil {
+		restutils.SendHTTPError(w, http.StatusBadRequest, errors.ErrVolNotFound.Error())
+		return
+	}
+
+	var req VolOptionRequest
+	if err := utils.GetJSONFromRequest(r, &req); err != nil {
+		restutils.SendHTTPError(w, http.StatusUnprocessableEntity, errors.ErrJSONParsingFailed.Error())
+		return
+	}
+
+	if !areOptionNamesValid(req.Options) {
```

Comment:
It'll be better here if you could log the invalid option, and also send it back to the user. Do the same also in volume create.
An explicit InvalidOptionError type will be useful here.

I'll hold merging this PR, till this is addressed.